### DISCUSSION TITLE
Would this PocketBase + Astro Multi-Tenancy Structure Work for a SaaS?

### DISCUSSION THREAD

**Discussion by kuwts** [pocketbase/pocketbase#6015] [Q&A]
Hi everyone! 👋

I'm working on a SaaS project using PB for the backend and Astro (with SSR) for the frontend. The idea is to create a platform for property management companies, where each company (tenant) gets their own templated website. These sites will include a homepage, an about page, and individual property pages. Each tenant's data will be isolated from others, following a multi-tenancy model.

Here's how I'm planning to structure things in PocketBase:

# Database Structure

1. `companies` **Collection (Tenants)**
   * Represents each property management company (tenant).
   * Fields:
      * `id`: Unique identifier for the company.
      * `name`: Name of the company.
      * `domain`: Subdomain or custom domain (e.g., `company1.saas.com`).
      * `logo`: Branding/logo file.
      * `homepage_content`: Custom homepage text or JSON (optional).
      * `created_at`, `updated_at`: Timestamps.
2. `properties` **Collection (Listings)**
   * Stores property listings for each company.
   * Fields:
      * `id`: Unique identifier.
      * `company_id`: Relational field linking to the `companies` collection.
      * `name`: Name of the property.
      * `description`: Description of the property.
      * `price`: Price or rental cost.
      * `images`: File uploads for photos.
      * `availability`: Boolean or enum for availability (e.g., available/sold/rented).
      * `created_at`, `updated_at`: Timestamps.
3. `users` **Collection (Tenant Users)**
   * Holds users for each company.
   * Fields:
      * `id`: Unique identifier.
      * `email`: User's email.
      * `password`: Encrypted password.
      * `role`: Enum (e.g., `admin`, `editor`, `viewer`) for permission levels.
      * `company_id`: Relational field linking to the `companies` collection.
      * `created_at`, `updated_at`: Timestamps.

# Workflow

1. **Tenant Isolation** Each company's data is scoped by its `company_id`. For example:
   * Fetching properties:
      * `jsCopy codeconst properties = await pb.collection('properties').getFullList(100, { filter: \`company\_id="${company.id}"\` });\`
   * Fetching users:
      * `jsCopy codeconst users = await pb.collection('users').getFullList(100, { filter: \`company\_id="${company.id}"\` });\`
2. **Dynamic Routing in Astro** I’ll dynamically detect the company from the request's domain or subdomain:
   * `jsCopy codeconst domain = new URL(request.url).hostname; const company = await pb.collection('companies').getFirstListItem(\`domain="${domain}"\`);\`
3. **Frontend Pages** The site structure is shared but populated dynamically for each tenant:
   * `bashCopy codesrc/pages/[domain]/`
      * `index.astro # Tenant-specific homepage`
      * `about.astro # About page`
      * `properties.astro # Property listings`
      * `properties/[id].astro # Single property details`

# Questions

1. Does this setup seem like a solid approach for multi-tenancy?
2. Are there any potential issues I should consider with scaling or isolating tenant data using this structure in PocketBase?
3. For those experienced with Astro or PocketBase, is there a better way to handle tenant-specific dynamic routing?
4. Is this structure secure enough to ensure data isolation between tenants?
5. Would there be a way to allow each company to use a custom domain somehow?

I’d really appreciate any feedback or advice! 🙏 Thanks in advance for your help!

**Comment by pnmcosta** [comment]
I have a Shopify App with a similar data structure, however all CRUD is auth protected (store user) and/or done internally, so I use https://pocketbase.io/docs/api-rules-and-filters/ to ensure it is only accessible to the store that owns it, e.g. in your case a rule on `properties` could be `@request.auth.id = company_id`

However this I think is something you've not covered in your design, how is auth going to occur? You should ensure that your PB implementation does not leak all properties as it would be open to abuse and scrapping.

PB, IMO is not great for machine-to-machine auth, and you may also want to read more about JS SSR caveats as described here https://github.com/pocketbase/pocketbase/discussions/5313

I can't comment on Astro specifically as I've never used it, but my app's frontend (on Shopify's Admin) is a React app (non-SSR) and my experience with integrating it with Pocketbase was seamless.

**Reply by pnmcosta** [reply]
One minor input on the structure, if multi user I would add to `properties`:

- `created_by`, `updated_by`: Relational fields linking to the `users` collection.

For auditing purposes.

**Reply by kuwts** [reply]
> I have a Shopify App with a similar data structure, however all CRUD is auth protected (store user) and/or done internally, so I use https://pocketbase.io/docs/api-rules-and-filters/ to ensure it is only accessible to the store that owns it, e.g. in your case a rule on `properties` could be `@request.auth.id = company_id`
> 
> However this I think is something you've not covered in your design, how is auth going to occur? You should ensure that your PB implementation does not leak all properties as it would be open to abuse and scrapping.
> 
> PB, IMO is not great for machine-to-machine auth, and you may also want to read more about JS SSR caveats as described here #5313
> 
> I can't comment on Astro specifically as I've never used it, but my app's frontend (on Shopify's Admin) is a React app (non-SSR) and my experience with integrating it with Pocketbase was seamless.

Thanks for the valid feedback! Didn't think about auth, I could potentially have a users table in a step above the companies table and that user would act as a sort of super admin potentially? However, if I created a users collection under each company ID would I be able to use Pocketbase's auth somehow?

**Reply by pnmcosta** [reply]
PB is not great for M2M auth, to avoid complexity (and have an acceptable level of risk) I would probably convert `companies` to a auth collection itself, and generate long lived tokens out of it for your Astro backend auth. It is not the ideal solution but should be workable, as long as you mitigate the risk of these long lived tokens being leaked somehow.

With v0.23 you can specific token duration for each collection rather than globally on previous versions.

**Reply by pnmcosta** [reply]
> However, if I created a users collection under each company ID would I be able to use Pocketbase's auth somehow?

Yes, you can create API rules on `properties` that could grant access to both `users` and `companies`. Relations and other collections are supported by the API rules.

**Reply by kuwts** [reply]
Thanks for the insight! Would M2M be necessary here, though? If the data is scoped to each company_id and each company has its own collection of users, wouldn’t PocketBase's regular auth be sufficient for isolation and security?

With API rules like @request.auth.company_id = company_id, I feel tenant data can already be effectively scoped and protected. Since each company's users are authenticated and manage their own data, wouldn’t adding M2M introduce unnecessary complexity for this use case?

Curious to hear your thoughts!

**Reply by pnmcosta** [reply]
Oh perhaps I misunderstood your initial requirements. 

If the properties aren't accessible to guests/visitor (only users) then perhaps M2M would not be needed. 

The reason I pointed to a m2m requiring was so that you could have isolation and guest access to view the Properties.

**Reply by kuwts** [reply]
Properties would be viewable on a public website belonging to each company (company_id) but wouldn't it be possible to just set open API rules for viewing, but then restricting the CRUD operations based on the company_id/users associated with it?

**Reply by pnmcosta** [reply]
That's the thing, if you set open api rules you  will be leaking all properties regardless of company. 
 
AFAIK using auth is the only to safely ensure tenant isolation.

**Reply by kuwts** [reply]
If you use Astro SSR you could get all the properties but then filter them to just those associated with that specific tenant, and that would all be done on the server.

**Reply by pnmcosta** [reply]
If you can ensure that the PB backend is never exposed publicly that would work. However, consider leveraging caching on Astros end, as you would essentially be fetching all properties to filter for a subset (by company) which will most definitely have an impact on performance.

**Reply by kuwts** [reply]
That's a fair point didn't think about filtering through all the properties, but maybe that wouldn't be needed either? As if you returned a company_id wouldn't you be able to simply return all the properties associated with that company_id without filtering through *all* the properties? For example:

```
// Fetch the user with expanded properties
    const user = await pb.collection('users').getOne(pb.authStore.model.id, {
      // Use expand to directly fetch related properties
      expand: 'company_id.properties'
    });

    // Directly access properties through the expanded relation
    const properties = user.expand['company_id'].expand?.properties || [];

    return {
      props: {
        properties,
        company: user.expand.company_id
      }
    };
```

This is just an example and not tested so not sure if it would work, but just what I thought it would look like in theory.

**Reply by pnmcosta** [reply]
That code does not look right to me, I don't think you would be able to expand properties like that. You would probably need to use back relations https://pocketbase.io/docs/working-with-relations/#back-relations in that example scenario. 

I guess the crucial bit would be to derive the `company_id` be it from a user or a guest visiting. There is a new feature on the latest version that I haven't explored yet, but my gut feeling is that it could be useful for your setup https://pocketbase.io/docs/authentication/#users-impersonation
